crates/subspace-networking/src/utils.rs

```diff
@@ -95,3 +97,148 @@ pub(crate) fn convert_multiaddresses(addresses: Vec<Multiaddr>) -> Vec<PeerAddre
+
+    /// The current outstanding permits
+    usage: usize,
+}
+
+impl SemState {
+    // Allocates a permit if available.
+    // Returns true if allocated, false otherwise.
+    fn alloc_one(&mut self) -> bool {
+        if self.usage < self.capacity {
+            self.usage += 1;
+            true
+        } else {
+            false
+        }
+    }
+
+    // Returns a free permit to the free pool.
+    // Returns true if any waiters need to be notified.
+    fn free_one(&mut self) -> bool {
+        let prev_is_full = self.is_full();
+        if let Some(dec) = self.usage.checked_sub(1) {
+            self.usage = dec;
+        } else {
+            panic!("SemState::free_one(): invalid free, state = {:?}", self);
+        }
+
+        // Notify if we did a full -> available transition.
+        prev_is_full && !self.is_full()
+    }
+
+    // Expands the max capacity by delta.
+    // Returns true if any waiters need to be notified.
+    fn expand(&mut self, delta: usize) -> bool {
+        let prev_is_full = self.is_full();
+        self.capacity += delta;
+
+        // Notify if we did a full -> available transition.
+        prev_is_full && !self.is_full()
+    }
+
+    // Shrinks the max capacity by delta.
+    fn shrink(&mut self, delta: usize) {
+        if let Some(dec) = self.capacity.checked_sub(delta) {
+            self.capacity = dec;
+        } else {
+            panic!("SemState::shrink(): invalid shrink, state = {:?}", self);
+        }
+    }
+
+    // Returns true if current usage exceeds capacity
+    fn is_full(&self) -> bool {
+        self.usage >= self.capacity
+    }
+}
+
+impl ResizableSemaphore {
+    pub(crate) fn new(capacity: usize) -> Self {
+        let shared = SemShared {
+            state: Mutex::new(SemState { capacity, usage: 0 }),
+            notify: Notify::new(),
+        };
+        Self(Arc::new(shared))
+    }
+
+    // Acquires a permit. Waits until a permit is available.
+    pub(crate) async fn acquire(&self) -> ResizableSemaphorePermit {
+        loop {
+            let wait = {
+                let mut state = self.0.state.lock().unwrap();
+                if state.alloc_one() {
+                    None
+                } else {
+                    // This needs to be done under the lock to avoid race.
+                    Some(self.0.notify.notified())
+                }
+            };
+
+            match wait {
+                Some(notified) => notified.await,
+                None => break,
+            }
+        }
+        ResizableSemaphorePermit(self.0.clone())
+    }
+
+    // Acquires a permit, doesn't wait for permits to be available.
+    // Currently used only for tests.
+    #[cfg(test)]
+    pub(crate) fn try_acquire(&self) -> Option<ResizableSemaphorePermit> {
+        let mut state = self.0.state.lock().unwrap();
+        if state.alloc_one() {
+            Some(ResizableSemaphorePermit(self.0.clone()))
+        } else {
+            None
+        }
+    }
```

Comment:
I'd rather move this to the test code itself. It will have access to private properties.

Reply:
I kept it here, as moving this to test makes the impl fragmented, making it hard to find the relevant functions. Also, wonder if we may have an use case for try_acquire() in future. In that case, we can get rid of the #[cfg(test)] directive